test/main.F90

```diff
@@ -1,84 +1,38 @@
-  type(test_diagnosis_test_t) test_diagnosis_test
-  type(vector_test_description_test_t) vector_test_description_test
-
-  type(command_line_t) command_line
-
-  integer :: passes=0, tests=0, skips=0
-
-  character(len=*), parameter :: usage = &
-    new_line('') // new_line('') // &
-    'Usage: fpm test -- [--help] | [--contains <substring>]' // &
-    new_line('') // new_line('') // &
-    'where square brackets ([]) denote optional arguments, a pipe (|) separates alternative arguments,' // new_line('') // &
-    'angular brackets (<>) denote a user-provided value, and passing a substring limits execution to' // new_line('') // &
-    'the tests with test subjects or test descriptions containing the user-specified substring.' // new_line('')
-
-  if (command_line%argument_present([character(len=len("--help"))::"--help","-h"])) stop usage
-
-  print "(a)", new_line("") // "Append '-- --help' or '-- -h' to your `fpm test` command to display usage information."
-
-  call assert_test%report(passes, tests, skips)
-  call bin_test%report(passes, tests, skips)
-  call formats_test%report(passes, tests, skips)
-  call string_test%report(passes, tests, skips)
-  call test_result_test%report(passes, tests, skips)
-  call test_description_test%report(passes, tests, skips)
-  call test_diagnosis_test%report(passes, tests, skips)
-  call vector_test_description_test%report(passes,tests, skips)
-
-  if (.not. GitHub_CI())  then
-    if (command_line%argument_present(["--test"])) then
-      call command_line_test%report(passes, tests, skips)
-    else
-      write(*,"(a)")  &
-      new_line("") // &
-      "To also test Julienne's command_line_t type, append the following to your fpm test command:" // &
-      new_line("") // &
-      "-- --test command_line_t --type"
-    end if
-  end if
-
-#if HAVE_MULTI_IMAGE_SUPPORT
-  if (this_image()==1) then
-#endif
-
-    print *
-    print '(*(a,:,g0))', "_________ In total, ",passes," of ",tests, " tests pass.  ", skips, " tests were skipped. _________"
+  use test_result_test_m             ,only :             test_result_test_t
 
-    if (passes + skips /= tests) error stop "Some executed tests failed."
+  implicit none
 
-#if HAVE_MULTI_IMAGE_SUPPORT
-  end if
+  ! Construct a test harness from an array of test fixtures, each of which is 
+  ! constructed from an invocation of a test_t child type's structure constructor:
+  associate(test_harness => test_harness_t([          &
+     test_fixture_t(                 assert_test_t()) &
+    ,test_fixture_t(                    bin_test_t()) &
+    ,test_fixture_t(                formats_test_t()) &
+    ,test_fixture_t(                 string_test_t()) &
+    ,test_fixture_t(       test_description_test_t()) &
+    ,test_fixture_t(         test_diagnosis_test_t()) &
+    ,test_fixture_t(            test_result_test_t()) &
+    ,test_fixture_t(           command_line_test_t()) &
+  ]))
+    call test_harness%report_results
+  end associate
```

Comment:
It looks like you've added a new user-facing `test_fixture_t` type, but I don't see any explicit documentation about it.

Should anything be added to the documentation about this type?

Reply:
Possibly.  My hope is that the subsequent PR that automatically generates a demonstration test suite eliminates the need for users to write this code other than copying and pasting one of the `test_fixture_t` lines and change the name of the test.

FYI, despite the fancy name, `test_fixture_t` currently serves as a workaround for not having containers.  The new type is a wrapper that allows for emulation of a polymorphic list of objects of heterogeneous type.  The fancy name is partly because it could be a more interesting type down the road if there's some need for more infrastructure in configuring tests, but for now, it's a simple wrapper type with a polymorphic component.

Reply:
@bonachea the new `test_harness_t` and `test_fixture_t` types are now mentioned in step 4 of the [Writing tests](https://github.com/BerkeleyLab/julienne/blob/auto-generate-driver/demo/README.md#writing-tests) section of `demo/README.md`.